### PR TITLE
543 sy 617 cesium panics when unable to parse channel key

### PR DESCRIPTION
**Pull request by LeonLiur**

### CODE REVIEW

File: cesium/channel_test.go [hunk 1]
```diff
@@ -74,6 +75,31 @@ var _ = Describe("Channel", func() {
...
+					Expect(f.Close()).To(Succeed())
+
+					db, err := cesium.Open("", cesium.WithFS(s))
+					Expect(err).ToNot(HaveOccurred())
+					Expect(db.Close()).To(Succeed())
+				})
+
+				It("Should turn parsable numeric folders into channels", func() {
+					s := MustSucceed(fs.Sub("sub"))
+					_, err := s.Sub("1")
```
Comment (emilbon99) [inline cesium/channel_test.go:91]: This is kind of concerning. This DB has no meta or index file, so we don't know the data type of the channel (along with other critical info). What will happen if we try to open a reader on this directory? I feel like this case should fail

**Review by emilbon99 — CHANGES_REQUESTED** [review]

File: cesium/channel_test.go (hunk 1 above)
Comment (LeonLiur) [reply cesium/channel_test.go:91]: The DB will create the meta file during `openUnaryOrVirtual` and will create `index.domain` and `counter.domain` files while opening the unary and domain DBs. So I think this is fine. If those files are corrupted, then I feel like DB creation should still succeed and internal errors should reflect those errors

Comment (LeonLiur) [reply cesium/channel_test.go:91]: NVM. On a second thought it doesn't make sense to create a meta file if we don't know the meta properties of that channel. I will make a check in `ReadOrCreate` meta that makes it fail if channel's properties are not set.

Comment (emilbon99) [reply cesium/channel_test.go:91]: This has been resolved?

File: cesium/internal/meta/meta.go [hunk 2]
```diff
@@ -29,16 +30,19 @@ func ReadOrCreate(fs xfs.FS, ch core.Channel, ecd binary.EncoderDecoder) (core.C
...
-				"[meta] - a channel is required when creating a new database",
-			)
-		}
-		return ch, Create(fs, ecd, ch)
+	if exists {
+		return Read(fs, ecd)
 	}
-	return Read(fs, ecd)
+
+	if ch.DataType == telem.UnknownT {
```
Comment (emilbon99) [inline cesium/internal/meta/meta.go:37]: We should validate other properties of the channel here no? Like checking that the key is not zero. Make sure we also check other properties that need to be well defined that I did not mention. Also, I feel like we should move this check into the `Create` function no?

File: cesium/open.go [hunk 3]
```diff
@@ -42,12 +42,14 @@ func Open(dirname string, opts ...Option) (*DB, error) {
...
 	for _, i := range info {
-		key := core.ChannelKey(lo.Must(strconv.Atoi(i.Name())))
-		if err != nil {
-			return nil, err
-		}
 		if i.IsDir() {
-			if err = _db.openVirtualOrUnary(Channel{Key: key}); err != nil {
+			key, err := strconv.Atoi(i.Name())
+			if err != nil {
+				_db.options.L.Error("failed parsing existing folder to channel key", zap.Error(err))
```
Comment (emilbon99) [inline cesium/open.go:48]: I don't think there's any point to call this variable `_db` anymore. Let's just rename to `db`

File: cesium/open.go [hunk 4]
```diff
@@ -42,12 +42,14 @@ func Open(dirname string, opts ...Option) (*DB, error) {
...
+				continue
+			}
+
+			if err = _db.openVirtualOrUnary(Channel{Key: ChannelKey(key)}); err != nil {
 				return nil, err
 			}
 		}
```
Comment (emilbon99) [inline cesium/open.go:55]: Maybe we should warn that there is an unknown file in the directory?

**Review by emilbon99 — CHANGES_REQUESTED** [review]

File: cesium/channel_test.go [hunk 5]
```diff
@@ -118,6 +119,108 @@ var _ = Describe("Channel", func() {
...
+					Expect(err).ToNot(HaveOccurred())
+					_, err = f.Write([]byte("heheheha"))
+					Expect(err).ToNot(HaveOccurred())
+					Expect(f.Close()).To(Succeed())
+
+					db, err = cesium.Open("", cesium.WithFS(s))
+					Expect(err).To(MatchError(ContainSubstring("error decoding meta file")))
+				})
+
+				It("Should error if the meta.json file got changed with impossible configurations", func() {
```
Comment (emilbon99) [inline cesium/channel_test.go:141]: This should be a table spec no?

File: cesium/channel_test.go [hunk 6]
```diff
@@ -118,6 +119,108 @@ var _ = Describe("Channel", func() {
...
+					Expect(ch.Key).To(Equal(key))
+					Expect(ch.Index).To(Equal(indexKey))
+					Expect(ch.DataType).To(Equal(telem.Int64T))
+
+					ch = MustSucceed(db.RetrieveChannel(ctx, indexKey))
+					Expect(ch).ToNot(BeNil())
+					Expect(ch.Key).To(Equal(indexKey))
+					Expect(ch.IsIndex).To(BeTrue())
+					Expect(ch.DataType).To(Equal(telem.TimeStampT))
+
```
Comment (emilbon99) [inline cesium/channel_test.go:211]: I would add some `By` clauses here just so its easier to identify where things go wrong.

File: cesium/internal/domain/index.go [hunk 7]
```diff
@@ -237,7 +237,7 @@ func (idx *index) read(f func()) {
 func (idx *index) persist(ctx context.Context, persistAtIndex int) error {
 	ctx, span := idx.T.Bench(ctx, "domain/index.persist")
 	defer span.End()
-	encoded := idx.indexPersist.encode(idx.persistHead, idx.mu.pointers)
```
Comment (emilbon99) [inline cesium/internal/domain/index.go:240]: Where do we update the persist head?

File: cesium/channel_test.go (hunk 5 above)
Comment (emilbon99) [reply cesium/channel_test.go:141]: Also, wouldn't it be possible to move these specs into the `meta` package instead?

**Review by emilbon99 — COMMENTED** [review]
Good to merge once you address my comments

**Pull request merged**